dataset/test/data_array_test.cpp

```diff
@@ -7,49 +7,63 @@
 #include "scipp/dataset/util.h"
 #include "scipp/variable/operations.h"
 
-#include "dataset_test_common.h"
 #include "test_macros.h"
 
 using namespace scipp;
-using namespace scipp::dataset;
 
-TEST(DataArrayTest, construct) {
-  DatasetFactory3D factory;
-  const auto dataset = factory.make();
+class DataArrayTest : public ::testing::Test {
+protected:
+  Variable data = makeVariable<double>(Values{1});
+  Variable coord = makeVariable<double>(Values{2});
+  Variable mask = makeVariable<bool>(Values{false});
+  Variable attr = makeVariable<double>(Values{3});
+};
+
+TEST_F(DataArrayTest, constructor_shares) {
+  DataArray a(data, {{Dim::X, coord}}, {{"mask", mask}}, {{Dim("attr"), attr}});
+  EXPECT_TRUE(a.data().is_same(data));
+  EXPECT_TRUE(a.coords()[Dim::X].is_same(coord));
+  EXPECT_TRUE(a.masks()["mask"].is_same(mask));
+  EXPECT_TRUE(a.attrs()[Dim("attr")].is_same(attr));
+}
 
-  DataArray array(dataset["data_xyz"]);
-  EXPECT_EQ(array, dataset["data_xyz"]);
-  // Comparison ignores the name, so this is tested separately.
-  EXPECT_EQ(array.name(), "data_xyz");
+TEST_F(DataArrayTest, copy_shares) {
+  const DataArray a(data, {{Dim::X, coord}}, {{"mask", mask}},
+                    {{Dim("attr"), attr}});
+  const DataArray b(a);
+  EXPECT_TRUE(a.data().is_same(b.data()));
+  EXPECT_TRUE(a.coords()[Dim::X].is_same(b.coords()[Dim::X]));
+  EXPECT_TRUE(a.masks()["mask"].is_same(b.masks()["mask"]));
+  EXPECT_TRUE(a.attrs()[Dim("attr")].is_same(b.attrs()[Dim("attr")]));
+  EXPECT_NE(&a.coords(), &b.coords());
+  EXPECT_NE(&a.masks(), &b.masks());
+  EXPECT_NE(&a.attrs(), &b.attrs());
 }
```

Comment:
Are this and the test above required given the ownership tests in Python? Or could the Python tests be removed in favour of these?

Reply:
Are you referring to the tests you added?

Reply:
Yes

Reply:
I think we generally try to have test mainly on the C++ side. Python tests are often omitted if the bindings are trivial. I think in the case of the ownership stuff the bindings are sufficiently complicated to justify re-testing?

Reply:
True. Let's keep both.